### PR TITLE
Add ability to spawn code notebooks from cBioPortal queries 

### PR DESCRIPTION
**Pull request by gautamsarawagi**
Jupyter-Notebook-lite in the frotnend for this [issue](https://github.com/cBioPortal/GSoC/issues/107)

Describe changes proposed in this pull request:
- Added the Jupyter Notebook Lite in the frontend code as follows: 
image
image
- Also sending the data of the studies in the **JupyterNotebook Page** , with which we can make prepare our models. These are as follows:
image
- Have remove the console although from the committed code.

## Checks
- [ ] Has tests or has a separate issue that describes the types of test that should be created. If no test is included it should explicitly be mentioned in the PR why there is no test.
- [ ] The commit log is comprehensible. It follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Is this PR adding logic based on one or more **clinical** attributes? If yes, please make sure validation for this attribute is also present in the data validation / data loading layers (in backend repo) and documented in [File-Formats Clinical data section](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#clinical-data)!

## Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [Giphy CAPTURE](https://giphy.com/apps/giphycapture) or [Peek](https://github.com/phw/peek)

## Notify reviewers
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). It can help to figure out who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them either through slack or by assigning them as a reviewer on the PR

### CODE REVIEW

**Comment by alisman** [comment]
Nice and quick work @gautamsarawagi !    Couple things:
1. can you make make a screenshot showing an simple example analysis of the passed data (instead of dummy data)
2. Right now we are passing the data in the url, which will most likely run up agains browser limits quickly.   JupiterLite offers a way of passing the data using post message I think, but we'll probably need to host the jupyterlite ourselves in our codebase, so we should go ahead and do that.

**Comment by gautamsarawagi** [comment]
@alisman You are absolutely right, We will need to host the JupyterLite ourself to pass the data to it. 

- I actually tried working with the passed data directly but that doesn't work actually. 
- But there is a way I have researched about it and got this [https://jupyterlite.readthedocs.io/en/stable/howto/configure/advanced/iframe.html#](doc). 

 Have got the direction now. Will keep you updated on progress.

**Comment by gautamsarawagi** [comment]
@alisman I have read all the docs in this last week. and tried out different cases: 
1. First I tried to run the algorithm on the passed data just as you suggested me to do and here is the result: 
image
**As you can see we are trying to access the data from the file and that is what now I am trying to do.**

2. I was trying to make **a connection between our hosted site with the Jupyter-lite instance** and tried  to execute this: 
[code](https://github.com/gautamsarawagi/communication_extension).
Got some progress in this code I was able to manage the themes from the hosted site as this [document](https://jupyterlite.readthedocs.io/en/stable/howto/configure/advanced/iframe.html) which is working like this: 
image

3. Similarly just as we can manage themes I am trying to manage the files also.

4. We will be passing the data from the site to the Jupyterlite notebook instance and update it there every time.

5. That would be really efficient according to me. For that, I found this [document](https://jupyterlab.readthedocs.io/en/latest/api/modules/codeeditor.html)

I would love to hear your thoughts on this and how I should move further.

**Comment by alisman** [comment]
@gautamsarawagi.  Looking good!  I like seeing the real data in the interface.
And yes it seems the extension is the right way to go to allow direct communication between portal and notebook.

Perhaps we can schedule a zoom session to meet eachother and try to talk about next steps. 

Thank you for your work!  Are you hoping to do GSOC this year?

--Aaron

**Comment by gautamsarawagi** [comment]
@alisman, thanks for the feedback! The extension does seem like the way to go for smooth communication between the portal and the notebook.

A Zoom session is a great idea to meet each other and discuss the next steps. Please do let me know your availability and  I'll make sure to accommodate accordingly.

And yes, I'm hoping to do **Google Summer of Code** this year! This project is right up my alley, so I'd love to contribute more if selected.

Appreciate you taking the time to review and provide helpful comments. Looking forward to our discussion and potential collaboration over the summer!

Regards,
Gautam

**Comment by gautamsarawagi** [comment]
Hey @alisman based on our discussion yesterday I came across this [extension](https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access) from the jupyterlab itself. And you were right. They are using **ServiceManager** API's. This might be a good resource. Also, I came across this [discussion](https://github.com/jupyterlite/jupyterlite/discussions/91) this gives a good understanding of how the files are stored internally.

I am working on this and will update you once the integration is successful.

**Comment by alisman** [comment]
@gautamsarawagi do you want me to look at anything in particular yet?

**Comment by gautamsarawagi** [comment]
Based on our discussion I came out with some **similar** available resources. So, actually wanted to share those.

**Comment by alisman** [comment]
@gautamsarawagi i think it would be good to include the notebook code itself in the PR.   We can figure out a way get the js loaded into an HTML page on the cbioportal.org domain.  THat will make progress and review quicker.  Ultimately, we may choose for security reasons to host the notebook on a different domain.  But the js itself can still live in this repository for convenience.

**Comment by gautamsarawagi** [comment]
Sure thats a good idea. It would be good to add the notebook code in the PR itself.

**Comment by gautamsarawagi** [comment]
@alisman I have added the notebook in the root directory of the code itself. and have also specified a Readme.md to use it. 

Although I was not able to reference the index.html from the root directory to the [JupyterNotebookTool.tsx](https://github.com/cBioPortal/cbioportal-frontend/blob/6684d511eeac8af20cdaba3a77a9d2f4e1dfed35/src/pages/staticPages/tools/oncoprinter/JupyterNotebookTool.tsx) : 

**I would also like to discuss this with you.**. I even tried adding [html-loaders](https://webpack.js.org/loaders/html-loader/) in the `webpack.config.js` file. But it did;t worked. 

So finally I deployed it for the use
image

If we can reference the notebook's index.html in the iframe then there won;t be the need for the deployment. Which will not even produce the security factors that are coming right now.

Also, I am now moving to my further work of working with the fileModification. 

Please do let me know if any further modifications are needed.

**Comment by alisman** [comment]
@gautamsarawagi i tested this and it looks good.  Here are some next steps:
- [ ] Give the output.csv a better name based on the query.  I believe the download option provides a name so you can get it from there
- [ ] Seed it with a python file (.pynb) which imports the file and does some very simple analysis on it.  Like, lists the column headers.

**Comment by gautamsarawagi** [comment]
Sure @alisman. First of all thanks for the feedback from your end. 

I will start working on showing a simple python file containing the file and running it.

Will share the progress with you.

**Comment by gautamsarawagi** [comment]
Hey @alisman 

- Updated the name of the file
- Attempting to auto-execute the ipynb file I am facing some issues with that but working on it. 
  - Till Now the file that needs to be auto-executedis getting auto-opened
  - Just need to figure out the way to execute the cells of it.

Will update you once the auto-execution part is done.

**PFA**: The changes in the extension created for the file-communication: [link](https://github.com/gautamsarawagi/file-communication-extension/blob/master/src/index.ts)

File: src/shared/components/oncoprint/ResultsViewOncoprint.tsx [hunk 1]
```diff
@@ -1115,14 +1139,82 @@ export default class ResultsViewOncoprint extends React.Component<
...
+                                    'sampleId',
+                                    'isHotspot',
+                                ];
+
+                                const allGenesMutationsCsv = convertToCSV(
+                                    allGenesMutations,
+                                    fieldsToKeep
+                                );
+
+                                this.jupyterFileContent = allGenesMutationsCsv;
```
Comment (alisman) [inline src/shared/components/oncoprint/ResultsViewOncoprint.tsx:1209]: you should delete ```jupyterFileContent``` this when the modal is closed

File: notebook/READEME.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (alisman) [inline notebook/READEME.md:1]: name of this file should be README.md (typo)

**Comment by alisman** [comment]
image

Add this HTML to button:

image

File: notebook/READEME.md (hunk 2 above)
Comment (gautamsarawagi) [reply notebook/READEME.md]: Corrected the typo.

**Comment by gautamsarawagi** [comment]
> image
> 
> Add this HTML to button:
> 
> image

Added the beta! text

**Pull request merged**